[FX] Normalize torch.nn.functional calls and standard Module calls

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51816 [FX] Normalize torch.nn.functional calls and standard Module calls**

Add a Transfomer that:

* Normalizes calls to `torch.nn.functional*` functions s.t. we use the function signature to move all args to kwargs in the correct positional order
* Does the same for stdlib `nn.Module` instances

Differential Revision: [D26290764](https://our.internmc.facebook.com/intern/diff/D26290764)

Commit:
[WIP][FX] Normalize torch.nn.functional calls

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Normalize torch.nn.functional calls"


Add a function that will take all `call_function` nodes with call targets in `torch.nn.functional`, bind the args and kwargs up to the function signature, and rewrite the arguments into a consistent ordered kwarg-only format.

Differential Revision: [D26290764](https://our.internmc.facebook.com/intern/diff/D26290764)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Normalize torch.nn.functional calls"


Add a function that will take all `call_function` nodes with call targets in `torch.nn.functional`, bind the args and kwargs up to the function signature, and rewrite the arguments into a consistent ordered kwarg-only format.

Differential Revision: [D26290764](https://our.internmc.facebook.com/intern/diff/D26290764)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Normalize torch.nn.functional calls and standard Module calls"


Add a function that will take all `call_function` nodes with call targets in `torch.nn.functional`, bind the args and kwargs up to the function signature, and rewrite the arguments into a consistent ordered kwarg-only format.

Differential Revision: [D26290764](https://our.internmc.facebook.com/intern/diff/D26290764)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Normalize torch.nn.functional calls and standard Module calls"


Add a function that will take all `call_function` nodes with call targets in `torch.nn.functional`, bind the args and kwargs up to the function signature, and rewrite the arguments into a consistent ordered kwarg-only format.

Differential Revision: [D26290764](https://our.internmc.facebook.com/intern/diff/D26290764)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Normalize torch.nn.functional calls and standard Module calls"


Add a function that will take all `call_function` nodes with call targets in `torch.nn.functional`, bind the args and kwargs up to the function signature, and rewrite the arguments into a consistent ordered kwarg-only format.

Differential Revision: [D26290764](https://our.internmc.facebook.com/intern/diff/D26290764)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Normalize torch.nn.functional calls and standard Module calls"


Add a function that will take all `call_function` nodes with call targets in `torch.nn.functional`, bind the args and kwargs up to the function signature, and rewrite the arguments into a consistent ordered kwarg-only format.

Differential Revision: [D26290764](https://our.internmc.facebook.com/intern/diff/D26290764)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Normalize torch.nn.functional calls and standard Module calls"


Add a function that will take all `call_function` nodes with call targets in `torch.nn.functional`, bind the args and kwargs up to the function signature, and rewrite the arguments into a consistent ordered kwarg-only format.

Differential Revision: [D26290764](https://our.internmc.facebook.com/intern/diff/D26290764)

[ghstack-poisoned]

Chillee commented on test/test_fx_experimental.py:
```diff
@@ -749,6 +751,97 @@ def test_subgraph_trivial_resnet(self):
+        m = resnet18()
+
+        class FunctionalTracer(torch.fx.Tracer):
+            def is_leaf_module(self, m: torch.nn.Module, module_qualified_name : str) -> bool:
```
So you want to decompose all the modules except for BatchNorm?

Chillee commented on test/test_fx_experimental.py:
```diff
@@ -749,6 +751,97 @@ def test_subgraph_trivial_resnet(self):
+
+        class FunctionalTracer(torch.fx.Tracer):
+            def is_leaf_module(self, m: torch.nn.Module, module_qualified_name : str) -> bool:
+                leaves = set([torch.nn.BatchNorm2d])
```
So you want to decompose all the modules except for BatchNorm?

jamesr66a replied:
For testing purposes I want to trace everything down to the functional level, but BatchNorm2d is currently not traceable due to control flow content (see https://fb.quip.com/EgT5A3DFrpaJ#IbQACAPoQ0n)

Chillee replied:
Might be worth adding a note.

Chillee commented on test/test_fx_experimental.py:
```diff
@@ -749,6 +751,97 @@ def test_subgraph_trivial_resnet(self):
+        torch.nn Module classes
+        """
+        for test_params in module_tests + new_module_tests:
+            if 'constructor' not in test_params:
```
I assume these are special case things, might be worth adding a brief comment to explain why.

jamesr66a replied:
Honestly not sure why this is the way it is. This is copypasta from `torch_nn`: https://github.com/pytorch/pytorch/blob/b2d8f0a431f8ce2c606ed25ac4d28287489c71bc/test/test_nn.py#L10063

What i'm trying to do here is just get instances + example inputs for all the stdlib `nn.Module` instances

Chillee commented on torch/fx/experimental/normalize.py:
```diff
@@ -0,0 +1,77 @@
+        if self.normalize_functionals and target.__module__ == 'torch.nn.functional':
+            target_for_analysis = target
+            if target in boolean_dispatched:
+                # HACK: `boolean_dispatch` as used in `torch.nn.functional` makes it so that we have
```
🤔 Why does this exist lol.

jamesr66a replied:
Because TorchScript (of course lol) https://github.com/pytorch/pytorch/pull/14425

Chillee commented on torch/fx/experimental/normalize.py:
```diff
@@ -0,0 +1,77 @@
+            classname = submod.__class__.__name__
+            if getattr(torch.nn, classname, None) == submod.__class__:
+                new_kwargs = self._args_kwargs_to_normalized_kwargs(submod.forward, args, kwargs)
+                if new_kwargs:
```
Should this be an `is None`?

jamesr66a replied:
No, `_args_kwargs_to_normalized_kwargs` returns `None` if this schema is not supported for normalization. We want to emit a `call_module` with `new_kwargs` if that is _not_ None

Chillee commented on torch/fx/experimental/normalize.py:
```diff
@@ -0,0 +1,77 @@
+            new_kwargs = self._args_kwargs_to_normalized_kwargs(target_for_analysis, args, kwargs)
+
+            if new_kwargs:
+                # FIXME: `target(**kwargs)` doesn't keep things specified as kwargs
```
Maybe also specify that this is a hack. I thought it was some sort of limitation of the current code.

jamesr66a replied:
This seems to be a limitation of `__torch_function__`. It seems to normalize args differently (i.e. it looks up schema and puts positional args in `args` and keyword args in `kwargs`). Not 100% sure though but I know `__torch_function__` does pull schemas

Chillee reviewed (approved):
LGTM

Commit:
Update on "[FX] Normalize torch.nn.functional calls and standard Module calls"


Add a Transfomer that:

* Normalizes calls to `torch.nn.functional*` functions s.t. we use the function signature to move all args to kwargs in the correct positional order
* Does the same for stdlib `nn.Module` instances

Differential Revision: [D26290764](https://our.internmc.facebook.com/intern/diff/D26290764)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize torch.nn.functional calls and standard Module calls"


Add a Transfomer that:

* Normalizes calls to `torch.nn.functional*` functions s.t. we use the function signature to move all args to kwargs in the correct positional order
* Does the same for stdlib `nn.Module` instances

Differential Revision: [D26290764](https://our.internmc.facebook.com/intern/diff/D26290764)

[ghstack-poisoned]
